### PR TITLE
GEODE-2817: Add new function security callback

### PR DESCRIPTION
**Pull request by jaredjstewart**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ x] Is your initial contribution a single, squashed commit?

- [ x] Does `gradlew build` run cleanly?

- [x ] Have you written or updated unit tests to verify your changes?

- [x ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/security/ResourcePermission.java [hunk 1]
```diff
@@ -14,6 +14,9 @@
  */
 package org.apache.geode.security;
 
+import java.util.List;
+import java.util.Set;
+
 import org.apache.commons.lang.StringUtils;
 import org.apache.geode.cache.Region;
 import org.apache.geode.management.internal.security.ResourceOperation;
```
Comment (PurelyApplied) [inline geode-core/src/main/java/org/apache/geode/security/ResourcePermission.java]: Both a nitpick and the fault of someone else, but the following files have unused or misordered imports:

```
geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/ExecuteFunctionCommand.java
geode-core/src/main/java/org/apache/geode/security/ResourcePermission.java
geode-core/src/test/java/org/apache/geode/internal/cache/OplogFlushTest.java
geode-core/src/test/java/org/apache/geode/internal/cache/OverflowOplogFlushTest.java
geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/CreateRegionCommandTest.java
geode-core/src/test/java/org/apache/geode/management/internal/security/SecurityServiceWithShiroIniIntegrationTest.java
```

File: geode-core/src/test/java/org/apache/geode/management/internal/security/CacheServerMBeanAuthorizationJUnitTest.java [hunk 2]
```diff
@@ -95,21 +95,21 @@ public void testNoAccess() throws Exception {
...
-        .hasMessageContaining(TestCommand.dataRead.toString());
+        .hasMessageContaining(ResourcePermissions.DATA_READ.toString());
     softly.assertThatThrownBy(() -> bean.fetchLoadProbe())
-        .hasMessageContaining(TestCommand.clusterRead.toString());
+        .hasMessageContaining(ResourcePermissions.CLUSTER_READ.toString());
     softly.assertThatThrownBy(() -> bean.getActiveCQCount())
-        .hasMessageContaining(TestCommand.clusterRead.toString());
+        .hasMessageContaining(ResourcePermissions.CLUSTER_READ.toString());
     softly.assertThatThrownBy(() -> bean.stopContinuousQuery("bar"))
         .hasMessageContaining(TestCommand.clusterManageQuery.toString());
```
Comment (PurelyApplied) [inline geode-core/src/test/java/org/apache/geode/management/internal/security/CacheServerMBeanAuthorizationJUnitTest.java]: Should `CLUSTER_MANAGE_QUERY` et al be added to `ResourcePermissions` and referenced as a common permission?  Or is just having a two-level resolution pre-defined sufficient for must use cases, allowing the three-level case to be explicitly defined at the point of use?

File: geode-core/src/test/java/org/apache/geode/test/dunit/rules/GfshShellConnectionRule.java [hunk 3]
```diff
@@ -173,6 +173,27 @@ public void connect(int port, PortType type, String... options) throws Exception
...
+    connectAndVerify(member.getJmxPort(), PortType.jmxManger);
+    runnable.run();
+    disconnect();
+    connectAndVerify(member.getHttpPort(), PortType.http);
+    runnable.run();
+    disconnect();
+  }
+
+  // when calling this method, the test needs to be in either geode-web or geode-assembly
+  public void secureExecuteOverJmxAndHttp(Member member, String username, String password,
```
Comment (PurelyApplied) [inline geode-core/src/test/java/org/apache/geode/test/dunit/rules/GfshShellConnectionRule.java]: These are excellent additions, but are currently unused.  `ExecuteFunctionCommandsWithSecurityDUnitTest` looks like a good candidate for application.

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/ExecuteFunctionCommand.java [hunk 4]
```diff
@@ -15,49 +15,43 @@
...
+
 import java.util.HashSet;
-import java.util.Iterator;
 import java.util.List;
+import java.util.Objects;
 import java.util.Set;
 import java.util.stream.Stream;
 
-import org.apache.commons.lang.BooleanUtils;
+import org.apache.shiro.session.Session;
```
Comment (pdxrunner) [inline geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/ExecuteFunctionCommand.java]: Looks like several unused imports

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/ExecuteFunctionCommand.java [hunk 5]
```diff
@@ -75,183 +69,64 @@ public Result executeFunction(
...
-    ResultCollector resultCollectorInstance = null;
-    Set<String> filters = new HashSet<>();
-    Execution execution;
-    if (functionId != null) {
-      functionId = functionId.trim();
-    }
-    if (onRegion != null) {
-      onRegion = onRegion.trim();
+
+    // when here, the options are already parsed and valided
```
Comment (pdxrunner) [inline geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/ExecuteFunctionCommand.java]: Spelling error in the comment 'valided' -> 'validated'

File: geode-core/src/main/java/org/apache/geode/cache/execute/Function.java [hunk 6]
```diff
@@ -106,11 +112,30 @@ public default boolean optimizeForWrite() {
...
-   * 
    * @since GemFire 6.5
    */
-  public default boolean isHA() {
+  default boolean isHA() {
     return true;
   }
 
+  /**
+   * returns the list of ResourcePermission this function requires.
```
Comment (pdxrunner) [inline geode-core/src/main/java/org/apache/geode/cache/execute/Function.java]: Bit of a nit-pick, but 'returns' should be capitalized.

File: geode-core/src/main/java/org/apache/geode/security/ResourcePermission.java [hunk 7]
```diff
@@ -14,6 +14,9 @@
  */
 package org.apache.geode.security;
 
+import java.util.List;
+import java.util.Set;
```
Comment (pdxrunner) [inline geode-core/src/main/java/org/apache/geode/security/ResourcePermission.java]: Unused imports.

File: geode-core/src/test/java/org/apache/geode/management/internal/security/CacheServerMBeanAuthorizationJUnitTest.java (hunk 2 above)
Comment (jaredjstewart) [reply geode-core/src/test/java/org/apache/geode/management/internal/security/CacheServerMBeanAuthorizationJUnitTest.java]: Let's tackle this question when we do the work to re-introduce String methods into SecurityService

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/result/ResultBuilder.java [hunk 8]
```diff
@@ -56,17 +52,17 @@
    * @param message Message to be shown to the user
    * @return Result for connection error
```
Comment (pdxrunner) [inline geode-core/src/main/java/org/apache/geode/management/internal/cli/result/ResultBuilder.java]: All the javadocs in here should now say the return type is CommandResult, not Result

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/functions/UserFunctionExecution.java [hunk 9]
```diff
@@ -29,115 +32,134 @@
...
+import org.apache.geode.internal.security.SecurityService;
 import org.apache.geode.management.internal.cli.GfshParser;
 import org.apache.geode.management.internal.cli.i18n.CliStrings;
 
 /**
- * 
  * @since GemFire 7.0
  */
 public class UserFunctionExecution implements Function, InternalEntity {
+  private static final Logger LOGGER = LogService.getLogger();
```
Comment (pdxrunner) [inline geode-core/src/main/java/org/apache/geode/management/internal/cli/functions/UserFunctionExecution.java]: Delete, LOGGER is not used. Also delete the corresponding import

**Pull request closed without merging**